# Evaluating Different ML Classification Algorithms For Reverse Vaccinology Prediction

## Table of Contents

- [Abstract](#abstract)
- [Dataset](#dataset)
- [Implementation details](#implementation-details)
- [Results](#results)
- [Citation](#citation)

## Abstract <a name="abstract"></a>

The research paper focusses on the utility and efficacy of various machine learning classifiers to identify potential vaccine candidates.

Reverse vaccinology utilises computational techniques to rationally design vaccines based on the information present in the genome of the pathogen. Bioinformatic tools screen the entire protein repetoire of each pathogen to select best vaccine candidates.
Various computational biology techniques have been designed to predict protein structure and function based on the protein sequence. The aim of this research paper is to predict which proteins could be vaccine candidates based on features derived from their amino acid sequence.

Here authors have accquired sequence information of various proteins from bacterial pathogens. This information has been processed by various sequence-based prediction tools to infer complex biological function of the proteins.
Multiple machine learning classifiers have been applied to the data to classify and predict which of the proteins have the potential to be vaccine candidates. Further each ML technique's ability to predict vaccine potential has been compared.

## Dataset <a name="dataset"></a>

This study utilised a previously assembled dataset BPAD200 developed by Heinson et al. [Heinson AI, Gunawardana Y, Moesker B, et al. Enhancing the Biological Relevance of Machine Learning Classifiers for Reverse Vaccinology. Int J Mol Sci. 2017;18(2):312. Published 2017 Feb 1. doi:10.3390/ijms18020312]

The dataset contained 200 Bacterial Protective Antigens (BPAs) and 200 non-BPAs. a BPA was defined as a bacterial protein that has led to significant protection (p <0.05) in an animal model following immunization and subsequent challenge with the bacterial pathogen.

The non-BPAs data was generated by randomly selecting proteins from the same bacterial species and same subcellular localization as each BPA. The filtering criteria for non-BPAs were:
- non-BPAs were discarded and resampled if they matched a BPA with more than 98% similarity using BLASTp
- non-BPAs were discarded and resampled if they matched to an already selected non-BPA with an E value > 10x10-3

The final selection of 200 BPAs and 200 non-BPAs were annotated to generate 525 features that could be used in Machine Learning. 34 protein annotation tools were utilised that described various biological properties of the protein such post-translational modification, methylation, amino acid percentage, MHC binding prediction etc.

Further details about how the features were generated can be accessed via the author's previous publication. [Heinson AI, Gunawardana Y, Moesker B, et al. Enhancing the Biological Relevance of Machine Learning Classifiers for Reverse Vaccinology. Int J Mol Sci. 2017;18(2):312. Published 2017 Feb 1. doi:10.3390/ijms18020312]

Preprocessed data (as provided by the author), which can be directly utilised for ML, has been provided in the Dataset folder. The datasets are :
- BPAD200- preprocessed data as given by the author comprising of 525 features
- BPAD200_Top10 - preprocessed data which includes only top 10 features from BPAD200; these features were derived by Greedy Backward Feature Elimination using SVM-RBF and mentioned by the author in previous literature
- BPAD200_Species - BPAD200 dataset with an additional feature that describes Bacterial species of each protein


## Implementation details <a name="implementation-details"></a>

Python 3.8 was used for machine learning implementation.

Scikit-Learn package was used to implement the following machine learning algorithms: Support Vector Machines, Logistic Regression and Random Forest. Results were evaluated as Area Under the Curve generated using a leave tenth out cross validation (LTOCV) on the entire dataset.

Greedy backward feature elimination (GBFE) was implemented to obtain an optimum top ten features for logistic regression from the entire dataset of 525 features. These top 10 features were already highlighted by the author in his previous publication. [Heinson AI, Gunawardana Y, Moesker B, et al. Enhancing the Biological Relevance of Machine Learning Classifiers for Reverse Vaccinology. Int J Mol Sci. 2017;18(2):312. Published 2017 Feb 1. doi:10.3390/ijms18020312]

To examine how efficient GBFE technique was in selecting informative features, permutation testing was conducted. 10 features were selected at random from the 525 features and used to classify BPAs and Non-BPAs. Classifier performances were evaluated using leave tenth out cross validation to generate an average AUC. This process was repeated 10,000 times to generate a frequency plot.

The authors also proposed a novel validation technique: Leave One Bacteria Out Validation (LOBOV). The dataset contains proteins from 40 different bacterial species. In LOBOV, every bacterial species in the dataset that had more than 4 examples of BPAs and non-BPAs combined were removed sequenctially from the training dataset and used as a testing dataset. 
This process helped the researchers understand whether ML techniques were capable of predicting BPAs from unknown or bacterial pathogens that were not included in the training dataset.


## Results <a name="results"></a>
A brief description of the results is included here. For further details and for the plots/graphs please check Results.pdf in the Results folder.

### A] Logistic Regression is as effective as Complex ML Algorithms in Predicting BPAs and Non-BPAs

The script for this procedure is MLAgos_Top10.py  The mean ROC AUC obtained for each of the algorithms is as follows:

- Logistic Regression:      0.7513333333333334
- Decision Tree Classifier: 0.6522083333333334
- Random Forest-10 trees:   0.6961666666666666
- Random Forest-100 trees:  0.7373333333333334
- Random Forest-1000 trees: 0.7384166666666666
- SVM Linear:               0.7464166666666666
- SVM Kernel:               0.7496666666666665
- SVM RBF:                  0.7396666666666668

It was observed that Logistic Regression obtains accuracies comparable to that of complex non-linear Support Vector Machine Classifiers. Only top 10 informative features from BPAD200 dataset were used in this analysis.
These were obtained by Greedy Backward Feature Elimination using SVM-RBF.

### B] Preprocessing improves Classification in Reverse Vaccinology

The script for this procedure is LR_Preprocessing.py

Here the authors have compared the results of ML techniques applied to the raw dataset versus a preprocessed dataset (where all features are scaled from -1 to 1). Since the author provided only the preprocessed dataset, the implementation included here is limited to the preprocessed dataset. Raw dataset has been excluded from this analysis.
However, based on the previous literature by the same authors, it could be possible to generate the raw dataset if needed.

The Mean ROC AUC obtained were :
- Logistic Regression on scaled BPAD200 dataset: 0.6765
- Logistic Regression on Top 10 features from BPAD200 dataset (features obtained by GBFE using SVM-RBF) : 0.7513333333333334
- Logistic Regression on Top 10 features from BPAD200 dataset (features obtained by GBFE using Logistic Regression): 0.8085

As hypothesised by the researchers, accuracy obtained by Logistic Regression improves when using the top 10 informative features from BPAD200 that were selected by Greedy Backward Feature Elimination using Logistic Regression.

### c] Greedy Backward Feature Elimination and Permutation Analysis

The scripts for this procedure is GBFE_LR.py

Analysis of Greedy Backward Feature Elimination technique showed that increase in features to predict BPAs from non-BPAs led to decrease in accuracy. Repeated iterations suggested that 'MBAAgl7_CorCount' was the most informative feature.
Similar results were obtained by the authors as well.  
The authors highlighted that 'MBAAgl7_CorCount' is a parameter that describes MHC class II epitope binding and is derived from protein annotation tool â€œGPS-MBA". The main function of major histocompatibility complex (MHC) class II molecules is to present processed antigens and are vital for for the initiation of the antigen-specific immune response.
The presence of this feature during prediction adds biological significance to the results.

Permutation analysis (which involved generation of 10,000 logistic classifiers with 10 random features from the BPAD200 dataset) revealed that GBFE strongly selects the most informative features even logistic regression classifier. 

### D] Leave One Bacteria Out Validation

The script for this procedure is LOBOV.py

Leave One Bacteria Out Validation (LOBOV) implementation revealed that machine learning techniques were capable of predicting potential vaccine candidates in bacterial species that were not included in the training dataset. However, dramatic fluctuations were observed in the ability to predict.
Therefore it was concluded that ML techniques in Reverse Vaccinology in BPAD200 dataset are more beneficial when predicting BPAs for certain species of bacterial pathogens when compared to other species.

Note: Differences in Results obtained by scripts in this repository versus those obtained by authors of the research papers could be attributed to the stochastic nature of algorithms implemented.


## Citation <a name="citation"></a>

Heinson AI, Ewing RM, Holloway JW, Woelk CH, Niranjan M. An evaluation of different classification algorithms for protein sequence-based reverse vaccinology prediction. PLoS One. 2019;14(12):e0226256. Published 2019 Dec 13. doi:10.1371/journal.pone.0226256
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6910663/ (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6910663/)

Heinson AI, Gunawardana Y, Moesker B, et al. Enhancing the Biological Relevance of Machine Learning Classifiers for Reverse Vaccinology. Int J Mol Sci.2017;18(2):312. Published 2017 Feb 1. doi:10.3390/ijms18020312
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5343848/ (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5343848/)
